torch/csrc/autograd/python_variable.cpp

```diff
@@ -116,80 +120,33 @@ static void THPVariable_dealloc(THPVariable* self)
+PyObject *THPVariable_pynew(PyTypeObject *type, PyObject *args, PyObject *kwargs)
 {
   HANDLE_TH_ERRORS
-  THPObjectPtr _data;
-  PyObject *data = nullptr;
-  PyObject *grad_fn = nullptr;
-  char is_volatile = 0;
-  char requires_grad = 0;
-  const char* name = nullptr;
-
-  const char *accepted_args[] = {"data", "requires_grad", "volatile", "_grad_fn", "name", nullptr};
-  if (!PyArg_ParseTupleAndKeywords(args, kwds, "|ObbOz", (char**)accepted_args,
-      &data, &requires_grad, &is_volatile, &grad_fn, &name))
-    return nullptr;
-
-  if (grad_fn == Py_None)
-    grad_fn = nullptr;
-
-  if (is_volatile) {
-    PyErr_WarnEx(PyExc_UserWarning, VOLATILE_WARNING, 1);
+  auto& default_type = torch::tensor::get_default_tensor_type();
+  if (default_type.is_cuda()) {
+    torch::utils::cuda_lazy_init();
   }
-
-  THPUtils_assert(!(is_volatile && requires_grad),
-          "Variable can't be volatile and require_grad at the same time!");
-  THPUtils_assert(!grad_fn || THPFunction_Check(grad_fn),
-          "Variable _grad_fn has to be a Function object or None, but got %s",
-          THPUtils_typename(grad_fn));
-  Tensor tensor;
-  if (!data || data == Py_None) {
-    // For legacy serialization code, create an empty tensor. This is also used
-    // by nn.Parameter() with no arguments.
-    auto var = torch::tensor::get_default_tensor_type().tensor();
-    tensor = static_cast<Variable&>(var).data();
-  } else if (THPVariable_Check(data)) {
-    tensor = ((THPVariable*)data)->cdata.data();
-  } else {
-    throw torch::TypeError("Variable data has to be a tensor, but got %s",
-        THPUtils_typename(data));
-  }
-
-  Variable var;
-  if (grad_fn) {
-    auto grad_fn_ = THPFunction_asFunction((THPFunction*)grad_fn);
-    Edge edge(grad_fn_, grad_fn_->bump_inputs());
-    var = make_variable(std::move(tensor), std::move(edge));
-  } else {
-    var = make_variable(std::move(tensor), requires_grad);
-  }
-
-  if (name) {
-    var.set_name(name);
-  }
-
-  return THPVariable_NewWithVar(type, std::move(var));
+  auto tensor = torch::utils::legacy_tensor_ctor(default_type, args, kwargs);
+  return THPVariable_NewWithVar(type, std::move(tensor));
   END_HANDLE_TH_ERRORS
 }
 
-int THPVariable_pyinit(PyObject *self, PyObject *args, PyObject *kwds)
-{
-  // Ensures that calls to Variable() and subclasses contain data argument.
-  // The 'data' argument is optional in __new__ to handle legacy serialized
-  // Variables.
-  PyObject *data;
-  PyObject *grad_fn = nullptr;
-  char is_volatile = 0;
-  char requires_grad = 0;
-  const char* name = nullptr;
-
-  const char *accepted_args[] = {"data", "requires_grad", "volatile", "_grad_fn", "name", nullptr};
-  if (!PyArg_ParseTupleAndKeywords(args, kwds, "|ObbOz", (char**)accepted_args,
-      &data, &requires_grad, &is_volatile, &grad_fn, &name))
-    return -1;
-
-  return 0;
+static PyObject* THPVariable_make_subclass(PyObject* _ignored, PyObject* args, PyObject* kwargs) {
```

Comment:
Is `_make_subclass` a hacky optimization to the constructor of `torch.Tensor`, that doesn't really care which class are you instantiating?

Reply:
It instantiates a subclass of torch.Tensor given an existing Tensor. It's used by `nn.Parameter` because that class now subclasses `torch.Tensor` instead of `Variable`.

We need a way to create a subclass without calling torch.Tensor() (i.e. `__new__`) because of the restrictions that Tensor.__new__ imposes.

For example,

```python
>>> torch.Tensor(torch.randn().double()) 
TypeError: expected Variable[torch.FloatTensor] (got Variable[torch.DoubleTensor])
```

But this works:

```python
>>> torch.nn.Parameter(torch.randn().double())
```